test/TestApps/PushNotificationsDemoApp/main.cpp

```diff
@@ -0,0 +1,146 @@
+{
+    // To obtain an AAD RemoteIdentifier for your app,
+    // follow the instructions on https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app
+    auto channelOperation = PushNotificationManager::CreateChannelAsync(
+        winrt::guid("ccd2ae3f-764f-4ae3-be45-9804761b28b2"));
+
+    // Setup the inprogress event handler
+    channelOperation.Progress(
+        [](auto&& sender, auto&& args)
+        {
+            if (args.status == PushNotificationChannelStatus::InProgress)
+            {
+                // This is basically a noop since it isn't really an error state
+                std::cout << "Channel request is in progress." << std::endl << std::endl;
+            }
+            else if (args.status == PushNotificationChannelStatus::InProgressRetry)
+            {
+                LOG_HR_MSG(
+                    args.extendedError,
+                    "The channel request is in back-off retry mode because of a retryable error! Expect delays in acquiring it. RetryCount = %d",
+                    args.retryCount);
+            }
+        });
+
+    auto result = co_await channelOperation;
+
+    if (result.Status() == PushNotificationChannelStatus::CompletedSuccess)
+    {
+        auto channelUri = result.Channel().Uri();
+
+        std::cout << "channelUri: " << winrt::to_string(channelUri.ToString()) << std::endl << std::endl;
+
+        auto channelExpiry = result.Channel().ExpirationTime();
+
+        // Register Push Event for Foreground
+        result.Channel().PushReceived([](const auto&, PushNotificationReceivedEventArgs const& args)
+            {
+                auto payload = args.Payload();
+
+                // Do stuff to process the raw payload
+                std::string payloadString(payload.begin(), payload.end());
+                std::cout << "Push notification content received from FOREGROUND: " << payloadString << std::endl << std::endl;
+                args.Handled(true);
+            });
+        // Caller's responsibility to keep the channel alive
+        co_return result.Channel();
+    }
+    else if (result.Status() == PushNotificationChannelStatus::CompletedFailure)
+    {
+        LOG_HR_MSG(result.ExtendedError(), "We hit a critical non-retryable error with channel request!");
+        co_return nullptr;
+    }
+    else
+    {
+        LOG_HR_MSG(result.ExtendedError(), "Some other failure occurred.");
+        co_return nullptr;
+    }
+
+};
+
+winrt::Microsoft::Windows::PushNotifications::PushNotificationChannel RequestChannel()
+{
+    auto task = RequestChannelAsync();
+    if (task.wait_for(std::chrono::seconds(300)) != AsyncStatus::Completed)
+    {
+        task.Cancel();
+        return nullptr;
+    }
+
+    auto result = task.GetResults();
+    return result;
+}
+
+int main()
+{
+    time_t ltime;
+    char buf[SIZE];
+
+    time(&ltime);
+    ctime_s(buf, sizeof buf, &ltime);
```

Comment:
C++ beats C:

```
#include <ctime>
...
auto now{ std::time(nullptr) };
auto localnow{ *std::localtime(&now) };
std::cout << "Blah blah: " << std::asctime(localnow) << std::endl;
```

Or replace `std::asctime()` with `std::put_time(&localnow, "%Y%m%d %H%M%S")` or whatever if you want more control over the format

https://en.cppreference.com/w/cpp/io/manip/put_time


Regardless, then there's no need to manage a raw buffer

Reply:
Actually there is no need for this now. Removing it. Thanks.